### PR TITLE
xds: implement xds_routing policy config generation and parsing

### PR DESCRIPTION
**Pull request by voidzcy**
In this PR, the xDS resolver generates a raw service config with load balancing config that supports all matchers in xds_routing config. This converts XdsClient's data structure `EnvoyProtoData.Route` to JSON format. Then the XdsRoutingConfig parser parses the JSON-format config back to a Java object. Note this convert-and-parse-back step is tedious and seemingly unnecessary, but this is how it works now as discussed in the design doc. In the future, we might eliminate this. Since `EnvoyProtoData.Route` and `XdsRoutingConfig` is somewhat similar in terms of its structure, we move the definition of matchers to a separate class so that the two can share the usage of types for parsed matchers.

Implementation for `XdsRoutingLoadBalancer` is temporarily deleted, as the definition of `XdsRoutingConfig` has changed completely so the routing policy needs to be re-implemeneted.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java [hunk 1]
```diff
@@ -97,77 +101,167 @@ public ConfigOrError parseLoadBalancingPolicyConfig(Map<String, ?> rawConfig) {
...
+      Map<String, ?> rawAction, LoadBalancerRegistry registry) {
+    List<LbConfig> childConfigCandidates = ServiceConfigUtil.unwrapLoadBalancingConfigList(
+        JsonUtil.getListOfObjects(rawAction, "childPolicy"));
+    if (childConfigCandidates == null || childConfigCandidates.isEmpty()) {
+      throw new RuntimeException("childPolicy not specified");
+    }
+    ConfigOrError selectedConfigOrError =
+        ServiceConfigUtil.selectLbPolicyFromList(childConfigCandidates, registry);
+    if (selectedConfigOrError.getError() != null) {
+      throw selectedConfigOrError.getError().asRuntimeException();
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java:144]: This wrapping and unwrapping create unnecessary stacktrace of exception. Some other places like
```java
} catch (RuntimeException e) {
  throw new RuntimeException("Failed to parse Route: " + e);
}
```
have the same problem

I feel the trick returning `StructOrErrorStatus<PolicySelection>` might be better.

File: xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java [hunk 2]
```diff
@@ -31,13 +33,15 @@
...
+import io.grpc.xds.RouteMatchers.HeaderMatcher;
+import io.grpc.xds.RouteMatchers.PathMatcher;
 import java.util.ArrayList;
-import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.Set;
+import java.util.regex.PatternSyntaxException;
```
Comment (voidzcy) [inline xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java:44]: This is wrong. It should be `import com.google.re2j.PatternSyntaxException;`

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java [hunk 3]
```diff
@@ -228,70 +230,126 @@ public void onError(Status error) {
...
+      }
+      if (suffixMatch != null) {
+        rawHeaderMatcherBuilder.put("suffixMatch", suffixMatch);
+      }
+      rawHeaderMatcherBuilder.put("invertMatch", headerMatcher.isInvertedMatch());
+      rawHeaderMatcherListBuilder.add(rawHeaderMatcherBuilder.build());
+    }
+    ImmutableList<?> rawHeaderMatchers = rawHeaderMatcherListBuilder.build();
+    if (!rawHeaderMatchers.isEmpty()) {
+      configRouteBuilder.put("headers", rawHeaderMatchers);
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsNameResolver.java:334]: The xds_routing config is not a mapping of the RouteMatch proto. It intends to be our own design, although the designer does just borrow the format and make them highly similar. In the design doc, the example xds_routing config does not include "headers" if not header matcher exists. So I followed that.

Also, the steps of converting what XdsClient receives to a raw xds_routing config and then converting it to a parsed xds_routing config is internal (and the designer says this double conversion is a temporary thing and will be eliminated in the short future), so as long as the config parser is able to parse the raw format correctly, whether it's an empty list or absence does not matter. (Btw, making the parser being able to parse absence seems to be more robust).

File: xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java (hunk 1 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java:143]: Slightly enhanced error propagation, threw away some stacktrace. Btw, the exception cases should ideally never happen. If it happens, the exception should have been surfaced to the Channel.

There is no perfect way to provide information about the error precisely, given that the JSON config is deeply nested and objects do not have keys that we can reference to. Unless we print out a tremendously long error message that locates the error level by level down, we are not able to point to the exact line of error in the JSON config. With that way, the error message would be something like:

```
Failed to parse xds_routing LB config: [a giant config JSON string for the whole config]
.....
Failed to parse HeaderMatcher: [a smaller but still big JSON object for the header matcher]
.....
Failed to parse Range: [a JSON object represents a Range]
....
```
This is hard to read and the effort it takes to find the malformed data is equal to proofread the whole JSON config.

The `StructOrErrorStatus` approach doesn't give better handling, it gives similar result but with verbose handling logic (at each step, always need to check if the enclosed is a struct or an error). Currently, we use a concise the error message that gives pretty much valuable hint for users to find the issue of the config.

**Pull request merged**